src/routes/registerPublicKeys.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  AuthorizedRequest,
+  Env,
+  ProfileWithId,
+  RegisterPublicKeyRequest,
+  RegisterPublicKeyResponse,
+} from '../types'
+import {
+  KnownError,
+  addProfilePublicKey,
+  getProfileFromPublicKey,
+  incrementProfileNonce,
+} from '../utils'
+import { verifyRequestBody } from '../utils/auth'
+
+export const registerPublicKeys = async (
+  request: AuthorizedRequest<RegisterPublicKeyRequest>,
+  env: Env
+) => {
+  const respond = (status: number, response: RegisterPublicKeyResponse) =>
+    new Response(JSON.stringify(response), {
+      status,
+    })
+
+  const {
+    auth: { publicKey },
+    publicKeys,
+  } = request.parsedBody.data
+
+  // Validate all keys inside.
+  try {
+    await Promise.all(publicKeys.map((key) => verifyRequestBody(key)))
+  } catch (err) {
+    if (err instanceof KnownError) {
+      return respond(err.statusCode, err.responseJson)
+    }
+
+    return respond(400, {
+      error:
+        'Failed to validate public keys: ' +
+        (err instanceof Error ? err.message : `${err}`),
+    })
+  }
+
+  // Validate all keys inside are different from each other.
+  const keySet = new Set(publicKeys.map((key) => key.data.auth.publicKey))
+  if (keySet.size !== publicKeys.length) {
+    return respond(400, {
+      error: 'Keys must be unique.',
+    })
+  }
+
+  // Validate all keys inside allowed this public key to register them.
+  if (publicKeys.some((key) => key.data.allow !== publicKey)) {
+    return respond(401, {
+      error: `Invalid allowed public key. Expected: ${publicKey}`,
+    })
+  }
+
+  // Get existing profile.
+  let profile: ProfileWithId | undefined
+  try {
+    profile = await getProfileFromPublicKey(env, publicKey)
+  } catch (err) {
+    console.error('Profile retrieval', err)
+
+    return respond(500, {
+      error:
+        'Failed to retrieve existing profile: ' +
+        (err instanceof Error ? err.message : `${err}`),
+    })
+  }
+
+  if (!profile) {
+    return respond(404, {
+      error: 'Profile not found.',
+    })
+  }
+
+  // Validate all nonces to prevent replay attacks.
+  if (
+    request.parsedBody.data.auth.nonce !== profile.nonce ||
```

Comment:
Should be `>` profile.nonce, right? Otherwise, there's a nonzero chance some earlier nonce could be replayed? Am I tripping?

Reply:
this makes sure the nonce submitted equals to nonce currently set in the profile. it doesn't allow anything less than or greater than. the nonce route returns the expected nonce that should be used in the next signature